### DISCUSSION TITLE
Why must `nodenv init -` be able to add `${NODENV_ROOT}/shims` to `$PATH` more than once?

### DISCUSSION THREAD

**Discussion by geoffreyvanwyk** [nodenv/nodenv#220] [Q&A] [answered]
There is a test for that at https://github.com/nodenv/nodenv/blob/master/test/init.bats#L87

Does it not clog up `$PATH` unnecessarily every time `~/.bashrc` is sourced?

**Comment by jasonkarns** [comment, marked as answer]
I don't have the source issues at hand (I believe they would be found under `rbenv`). The crux boiled down to: `nodenv init` should not make assumptions about presumed intentions when it runs. Its job is to ensure the nodenv shims are first in `PATH` (at the time nodenv-init is run) to ensure that nodenv's shims take precedence over whatever is already in `PATH`.

At one point, it tried to be overly clever and remove the existing `PATH` entries. However, that was found to be brittle and unsafe in environments where `PATH` was set by other entities prior to nodenv. (distro-provided defaults, or sysadmin-controlled init scripts, for example) And it removed control from the users who perhaps didn't want PATH to be mutated beyond the prefixing.

If a user's `PATH` ends up with the shims in it twice, the recommendation would be to fix whatever is (likely erroneously) adding them the first time prior to nodenv-init being invoked.

**Reply by jasonkarns** [reply]
Related conversation: https://github.com/rbenv/rbenv/issues/1573
